pg_search/src/postgres/storage/fsm.rs

```diff
@@ -34,413 +35,1263 @@ enum FSMBlockKind {
-///
-/// ```text
-/// [kind (4 bytes)] [empty (1 byte)] [ [blockno (4 bytes)] [xid (4 bytes)] ][ ... ] (up to MAX_ENTRIES_PER_PAGE)
-/// ```
-///
-/// And blocks are linked together with the `next_blockno` field in the [`BM25PageSpecialData`].
-///
-#[derive(Debug)]
-pub struct FreeSpaceManager {
-    start_blockno: pg_sys::BlockNumber,
-}
+pub trait FreeSpaceManager {
+    /// Create a new [`V1FSM`] in the block storage of the specified `indexrel`.
+    unsafe fn create(indexrel: &PgSearchRelation) -> pg_sys::BlockNumber;
 
-impl FreeSpaceManager {
-    /// Create a new [`FreeSpaceManager`] in the block storage of the specified `indexrel`.
-    pub unsafe fn create(indexrel: &PgSearchRelation) -> pg_sys::BlockNumber {
-        let mut new_buffer = init_new_buffer(indexrel);
-        let mut page = new_buffer.page_mut();
-        *page.contents_mut::<FSMBlock>() = FSMBlock::default();
-        new_buffer.number()
-    }
-
-    /// Open an existing [`FreeSpaceManager`] which is rooted at the specified starting block number.
-    pub fn open(start_blockno: pg_sys::BlockNumber) -> Self {
-        Self { start_blockno }
-    }
+    /// Open an existing [`V1FSM`] which is rooted at the specified starting block number.
+    fn open(start_blockno: pg_sys::BlockNumber) -> Self;
 
     /// Retrieve a single recyclable [`pg_sys::BlockNumber`], which can be acquired and re-initialized.
     ///
     /// Returns `None` if no recyclable blocks are available.
     ///
-    /// Upon return, the block is removed from the [`FreeSpaceManager`]'s control.  It is the caller's
+    /// Upon return, the block is removed from the [`V1FSM`]'s control.  It is the caller's
     /// responsibility to ensure the block is properly used, or else it will be lost forever as
     /// dead space in the underlying relation.
-    pub fn pop(&mut self, bman: &mut BufferManager) -> Option<pg_sys::BlockNumber> {
-        let xid_horizon =
-            unsafe { pg_sys::GetCurrentTransactionIdIfAny().max(pg_sys::FirstNormalTransactionId) };
-        let mut blockno = self.start_blockno;
-        loop {
-            if blockno == pg_sys::InvalidBlockNumber {
-                return None;
-            }
+    fn pop(&mut self, bman: &mut BufferManager) -> Option<pg_sys::BlockNumber>;
 
-            let buffer = bman.get_buffer(blockno);
-            let page = buffer.page();
+    /// Drain `n` recyclable blocks from this [`V1FSM`] instance, using the specified
+    /// [`BufferManager`] for underlying disk access.
+    ///
+    /// As [`pg_sys::BlockNumber`]s are yielded from the returned iterator, they are removed from the
+    /// FSM.  The returned iterator will never return more than `n`, but it could return fewer.
+    ///
+    /// It is the caller's responsibility to ensure each yielded block is properly used, or else it will
+    /// be lost forever as dead space in the underlying relation.  Unyielded blocks are unaffected.
+    fn drain(
```

Comment:
Not the subject of this review, but using drain in more places would probably be good for contention:
* `BlockList::finish` has an in-memory structure that it knows the length of, and could drain N blocks to store it in.
* `LinkedItemList::atomically` is cloning a known-length list
* `LinkedBytesListWriter::write` could "read ahead" and allocate additional space with drain if had a hint from the writer.

Reply:
yeah, the first two you noted could be cleaned up in a future commit.  `atomically` really should just go into a `Vec<SegmentMetaEntry>` first, anyways.

Reading ahead/allocating space ahead of time comes with the risk of those blocks not actually being used and then going *poof*.

Reply:
> Reading ahead/allocating space ahead of time comes with the risk of those blocks not actually being used and then going poof.

It does unless there is a `Drop` handler that returns them, yea. For really big bulk writes, I expect it would be worthwhile.